feat: add C ndarray interface and refactor implementation for `stats/base/dcumax`

aayush0325 opened the pull request:
Resolves None

## Description

> What is the purpose of this pull request?

This pull request:

-   adds C ndarray interface.
-   refactors implementation wherever needed.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves None.

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

aayush0325 committed:
feat: add C ndarray interface and refactor implementation

aayush0325 committed:
fix: bug in ndarray.native.js fixed

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

aayush0325 committed:
chore: fix linting errors

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

aayush0325 committed:
docs: update docs to follow latest project conventions

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

aayush0325 commented:
ready for initial review @kgryte!

aayush0325 committed:
feat: update js examples

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/stats/base/dcumax/README.md:
```diff
@@ -166,6 +165,120 @@ console.log( y );
+#### stdlib_strided_dcumax( N, \*X, strideX, \*Y, strideY )
+
+Computes the cumulative maximum of double-precision floating-point strided array elements.
+
```
Missing a C example.

aayush0325 replied:
What should i write here for a void function? had some confusions about that so i left it for now @kgryte

kgryte replied:
See, e.g., https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/blas/base/dcopy.

kgryte replied:
Also, rather than omit entirely, better to insert a TODO, so that reviewers know that something is missing. Otherwise, can be easy to miss.

kgryte commented on lib/node_modules/@stdlib/stats/base/dcumax/README.md:
```diff
@@ -141,20 +141,19 @@ dcumax.ndarray( 4, x, 2, 1, y, -1, y.length-1 );
 var Float64Array = require( '@stdlib/array/float64' );
 var dcumax = require( '@stdlib/stats/base/dcumax' );
 
 var y;
 var x;
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/stats/base/dcumax/README.md:
```diff
@@ -141,20 +141,19 @@ dcumax.ndarray( 4, x, 2, 1, y, -1, y.length-1 );
 var y;
 var x;
-var i;
 
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/stats/base/dcumax/README.md:
```diff
@@ -141,20 +141,19 @@ dcumax.ndarray( 4, x, 2, 1, y, -1, y.length-1 );
 var x;
-var i;
 
 x = new Float64Array( 10 );
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/stats/base/dcumax/README.md:
```diff
@@ -141,20 +141,19 @@ dcumax.ndarray( 4, x, 2, 1, y, -1, y.length-1 );
-var i;
 
 x = new Float64Array( 10 );
 y = new Float64Array( x.length );
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/stats/base/dcumax/README.md:
```diff
@@ -141,20 +141,19 @@ dcumax.ndarray( 4, x, 2, 1, y, -1, y.length-1 );
-for ( i = 0; i < x.length; i++ ) {
-    x[ i ] = round( randu()*100.0 );
-}
+
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/stats/base/dcumax/README.md:
```diff
@@ -141,20 +141,19 @@ dcumax.ndarray( 4, x, 2, 1, y, -1, y.length-1 );
-    x[ i ] = round( randu()*100.0 );
-}
+
+x = discreteUniform( 10, -50, 50, {
```
```suggestion
var x = discreteUniform( 10, -50, 50, {
```

kgryte commented on lib/node_modules/@stdlib/stats/base/dcumax/README.md:
```diff
@@ -141,20 +141,19 @@ dcumax.ndarray( 4, x, 2, 1, y, -1, y.length-1 );
+    'dtype': 'float64'
+});
 console.log( x );
 console.log( y );
```
```suggestion

y = new Float64Array( x.length );
console.log( y );
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/stats/base/dcumax/README.md:
```diff
@@ -166,6 +161,120 @@ console.log( y );
+#### stdlib_strided_dmax_ndarray( N, \*X, strideX, offsetX, \*Y, strideY, offsetY )
+
+Computes the cumulative maximum of double-precision floating-point strided array elements using alternative indexing semantics.
+
```
Missing a C example.

kgryte commented on lib/node_modules/@stdlib/stats/base/dcumax/benchmark/benchmark.js:
```diff
@@ -40,14 +47,9 @@ var dcumax = require( './../lib/dcumax.js' );
 */
 function createBenchmark( len ) {
 	var y;
```
Don't do this sort of mixed declaration and assignment. In this case, it is fine to do

```
var x = uniform(...);
var y = new Float64Array( len );
```

For more complex scenarios, declare first and then assign later.

kgryte commented on lib/node_modules/@stdlib/stats/base/dcumax/benchmark/benchmark.native.js:
```diff
@@ -48,15 +51,10 @@ var opts = {
 */
 function createBenchmark( len ) {
-	var x;
 	var y;
```
Same comment.

kgryte commented on lib/node_modules/@stdlib/stats/base/dcumax/benchmark/benchmark.ndarray.js:
```diff
@@ -39,15 +46,10 @@ var dcumax = require( './../lib/ndarray.js' );
 */
 function createBenchmark( len ) {
-	var x;
 	var y;
```
Same comment.

kgryte commented on lib/node_modules/@stdlib/stats/base/dcumax/benchmark/benchmark.ndarray.native.js:
```diff
@@ -48,15 +51,10 @@ var opts = {
 */
 function createBenchmark( len ) {
-	var x;
 	var y;
```
Same comment.

kgryte commented on lib/node_modules/@stdlib/stats/base/dcumax/docs/repl.txt:
```diff
@@ -3,8 +3,8 @@
 
-    The `N` and `stride` parameters determine which elements in `x` and `y` are
-    accessed at runtime.
+    The `N` and `stride` parameters determine which elements in the strided
```
```suggestion
    The `N` and stride parameters determine which elements in the strided
```

"arrays" can likely now fit on the previous line. Meaning, you'll likely need to rewrap the text.

kgryte commented on lib/node_modules/@stdlib/stats/base/dcumax/docs/repl.txt:
```diff
@@ -44,21 +44,20 @@
     // Using `N` and `stride` parameters:
```
```suggestion
    // Using `N` and stride parameters:
```

kgryte commented on lib/node_modules/@stdlib/stats/base/dcumax/examples/index.js:
```diff
@@ -18,20 +18,19 @@
 var Float64Array = require( '@stdlib/array/float64' );
 var dcumax = require( './../lib' );
 
 var y;
```
Need to update to be the same as the README.

kgryte commented on lib/node_modules/@stdlib/stats/base/dcumax/lib/dcumax.js:
```diff
@@ -41,58 +41,17 @@ var isPositiveZero = require( '@stdlib/math/base/assert/is-positive-zero' );
-		}
-	}
-	return y;
+	ox = stride2offset( N, strideX );
```
Feel free to declare and assign here, rather than declare and then assign later.

kgryte commented on lib/node_modules/@stdlib/stats/base/dcumax/src/main.c:
```diff
@@ -30,26 +31,35 @@
 * @param strideY  Y stride length
 */
-void stdlib_strided_dcumax( const int64_t N, const double *X, const int64_t strideX, double *Y, const int64_t strideY ) {
+void API_SUFFIX( stdlib_strided_dcumax ) ( const CBLAS_INT N, const double *X, const CBLAS_INT strideX, double *Y, const CBLAS_INT strideY ) {
```
```suggestion
void API_SUFFIX(stdlib_strided_dcumax)( const CBLAS_INT N, const double *X, const CBLAS_INT strideX, double *Y, const CBLAS_INT strideY ) {
```

Exception to the rule due to conventions elsewhere.

kgryte commented on lib/node_modules/@stdlib/stats/base/dcumax/src/main.c:
```diff
@@ -30,26 +31,35 @@
+* @param strideX  X stride length
+* @param Y        output array
+* @param strideY  Y stride length
+* @param offsetX  starting index for X
```
These parameter descriptions are in the wrong order and don't match the signature.

kgryte commented on lib/node_modules/@stdlib/stats/base/dcumax/src/main.c:
```diff
@@ -30,26 +31,35 @@
+* @param offsetX  starting index for X
+* @param offsetY  starting index for Y
+*/
+void API_SUFFIX( stdlib_strided_dcumax_ndarray ) ( const CBLAS_INT N, const double *X, const CBLAS_INT strideX, const CBLAS_INT offsetX, double *Y, const CBLAS_INT strideY, const CBLAS_INT offsetY ) {
```
```suggestion
void API_SUFFIX(stdlib_strided_dcumax_ndarray)( const CBLAS_INT N, const double *X, const CBLAS_INT strideX, const CBLAS_INT offsetX, double *Y, const CBLAS_INT strideY, const CBLAS_INT offsetY ) {
```

kgryte commented on lib/node_modules/@stdlib/stats/base/dcumax/test/test.ndarray.js:
```diff
@@ -330,9 +318,8 @@ tape( 'the function supports complex access patterns', function test( t ) {
-	N = 3;
 
-	dcumax( N, x, 2, 0, y, -1, 2 );
+	dcumax( 6, x, 2, 0, y, -1, 2 );
```
```suggestion
	dcumax( 3, x, 2, 0, y, -1, 2 );
```

kgryte commented on lib/node_modules/@stdlib/stats/base/dcumax/README.md:
````diff
@@ -166,6 +161,120 @@ console.log( y );
+```c
+#include "stdlib/stats/base/dcumax.h"
+#include <stdio.h>
+#include <inttypes.h>
````
This header is not needed.

kgryte commented on lib/node_modules/@stdlib/stats/base/dcumax/examples/c/example.c:
```diff
@@ -17,27 +17,26 @@
 #include "stdlib/stats/base/dcumax.h"
-#include <stdint.h>
 #include <stdio.h>
 #include <inttypes.h>
```
This header is not needed, as `PRId64` was removed.

kgryte reviewed (changes requested):
Left an initial round of comments.

aayush0325 commented:
> Left an initial round of comments.

i'll push the required fixes soon, thanks @kgryte!

aayush0325 committed:
chore: stuff from code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

aayush0325 committed:
chore: stuff from code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

aayush0325 committed:
chore: update examples to match README.md

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: passed
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

aayush0325 committed:
chore: incorrect function name in README.md

aayush0325 committed:
chore: typo in README.md fix

aayush0325 committed:
chore: return statement in newline

aayush0325 committed:
chore: fix incorrect spacing

aayush0325 committed:
chore: update params description in repl.txt

kgryte committed:
chore: clean-up

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

kgryte reviewed (approved):
This now LGTM. As a heads-up, I made a couple of small changes that may be applicable to other similar PRs, @aayush0325.

aayush0325 commented:
> This now LGTM. As a heads-up, I made a couple of small changes that may be applicable to other similar PRs, @aayush0325.

I'll take care of that, thanks @kgryte!
